packages/jest-config/src/readConfigFileAndSetRootDir.ts

```diff
@@ -94,9 +95,40 @@ const loadTSConfigFile = async (
     });
   } catch (e: any) {
     if (e.code === 'MODULE_NOT_FOUND') {
-      throw new Error(
-        `Jest: 'ts-node' is required for the TypeScript configuration files. Make sure it is installed\nError: ${e.message}`,
-      );
+      let esbuild: typeof build;
+
+      try {
+        esbuild = require('esbuild').build;
+      } catch (error: any) {
+        if (error.code === 'MODULE_NOT_FOUND') {
+          throw new Error(
+            `Jest: 'ts-node' or 'esbuild' is required for the TypeScript configuration files. Make sure it is installed\nError: ${e.message}`,
+          );
+        }
+
+        throw error;
+      }
+
+      const outfile = 'node_modules/jest-config/_jest.config.js';
+
+      await esbuild({
+        // By bundling we add support for importing stuff in the config file.
+        bundle: true,
+        entryPoints: [configPath],
+        format: 'cjs',
+        outfile,
+      });
+
+      let configObject = interopRequireDefault(require(outfile)).default;
+
+      await require('fs').unlink(outfile);
+
+      // In case the config is a function which imports more Typescript code
+      if (typeof configObject === 'function') {
+        configObject = await configObject();
+      }
```

Comment:
My understanding is that this won't work for `esbuild`, since it's not hooked into `require` like `ts-node/register` does; this could mean this is a non-starter since it's reasonable for people to assume importing would work.

I think using [`esbuild-register`](https://github.com/egoist/esbuild-register) would be better for this.

Reply:
btw my other comments become a bit moot I think if you switch to using `esbuild-register`.

Reply:
Indeed dynamic importing wouldn't work because it won't bundle that. What use cases are there for dynamic importing in config files?

I think `esbuild-register` won't work for this because it doesn't allow the overriding of the file format to CJS.  But I can try.

Reply:
Now using `esbuild-register` 😁